### PR TITLE
Deprecate registering plugins by class

### LINKED ISSUE #5698: `distributed.Client.register_worker_plugin` silently discards `**kwargs`
**What happened**:
An issue was [raised in slack](https://coiled-users.slack.com/archives/C0195GJKQ1G/p1643038235023300) regarding `UploadPlugin` failing to update the `sys.path` or restart workers. It was clear from the example code that `update_path=True` and `restart=True`, however these `kwargs` are actually ignored because a plugin instance was passed instead of a class.

**Minimal Complete Verifiable Example**:

```python
import asyncio
import pathlib
import tempfile

import distributed
from distributed import Nanny
from distributed.diagnostics.plugin import UploadDirectory


def run_demo_module():
    import demo_module

    return demo_module.demo_function()


async def amain():
    async with distributed.Client(
        processes=True, security=True, asynchronous=True
    ) as client:
        with tempfile.TemporaryDirectory() as tmpdir_fn:
            tmpdir = pathlib.Path(tmpdir_fn)
            (tmpdir / "demo_module.py").write_bytes(
                b'def demo_function(): return "hello"\n'
            )

            await client.register_worker_plugin(
                UploadDirectory(tmpdir),
                update_path=True,
                restart=True,
                nanny=True,
            )

        assert await client.submit(run_demo_module) == "hello"


if __name__ == "__main__":
    asyncio.run(amain())
```

results in:

```pytb
distributed.worker - WARNING - Compute Failed
Function:  run_demo_module
args:      ()
kwargs:    {}
Exception: 'ModuleNotFoundError("No module named \'demo_module\'")'

Traceback (most recent call last):
  File "demo.py", line 37, in <module>
    asyncio.run(amain())
  File "/home/graingert/miniconda3/envs/my_env/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/home/graingert/miniconda3/envs/my_env/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "demo.py", line 32, in amain
    assert await client.submit(run_demo_module) == "hello"
  File "/home/graingert/miniconda3/envs/my_env/lib/python3.8/site-packages/distributed/client.py", line 245, in _result
    raise exc.with_traceback(tb)
  File "demo.py", line 10, in run_demo_module
    import demo_module
ModuleNotFoundError: No module named 'demo_module'
```

for this case the fix is to apply the kwargs directly to `UploadDirectory`:
```diff
@@ -24,9 +24,11 @@
             )
 
             await client.register_worker_plugin(
-                UploadDirectory(tmpdir),
-                update_path=True,
-                restart=True,
+                UploadDirectory(
+                    tmpdir,
+                    update_path=True,
+                    restart=True,
+                ),
                 nanny=True,
             )
```

### PR DESCRIPTION
**Pull request by graingert**
- [x] Closes #5698
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: distributed/client.py [hunk 1]
```diff
@@ -4601,9 +4601,6 @@ def register_worker_plugin(self, plugin=None, name=None, nanny=None, **kwargs):
             If plugin has no name attribute a random name is used.
         nanny : bool, optional
             Whether to register the plugin with workers or nannies.
-        **kwargs : optional
-            If you pass a class as the plugin, instead of a class instance, then the
-            class will be instantiated with any extra keyword arguments.
```
Comment (fjetter) [inline distributed/client.py:4606]: If we kill this doc string we should be more specific for the `plugin` argument docs now, e.g.

`The initialized plugin to register.`

File: distributed/client.py [hunk 2]
```diff
@@ -4638,7 +4635,12 @@ class will be instantiated with any extra keyword arguments.
         distributed.WorkerPlugin
         unregister_worker_plugin
         """
-        if isinstance(plugin, type):
+        if isinstance(plugin, type) or kwargs:
+            warnings.warn(
+                "passing plugin classes or kwargs is deprecated; "
+                "pass a constructed plugin instance instead",
+                category=FutureWarning,
+            )
```
Comment (graingert) [inline distributed/client.py:4643]:
```suggestion
        if isinstance(plugin, type) or kwargs:
            warnings.warn(
                "passing plugin classes or kwargs is deprecated; "
                "pass a constructed plugin instance instead",
                category=FutureWarning,
            )

        if isinstance(plugin, type):
```

Comment (graingert) [inline distributed/client.py:4643]: this PR also changes the behaviour to *always* bind the kwargs to the plugin if any are passed, if that's not desirable something like this could be done instead:

```suggestion
        if isinstance(plugin, type) or kwargs:
            warnings.warn(
                "passing plugin classes or kwargs is deprecated; "
                "pass a constructed plugin instance instead",
                category=FutureWarning,
            )
        if isinstance(plugin, type):
```

Comment (bryanwweber) [inline distributed/client.py:4643]: Sorry if I missed the discussion somewhere, is there a reason to deprecate this feature? It seems useful to me and many other plugin type packages allow passing a class as the plugin so it may be something users expect. Certainly an improved warning message or fixing the broken behavior is warranted.

Comment (graingert) [reply distributed/client.py:4643]: there's a similar PR here: https://github.com/dask/distributed/pull/5394/files#diff-bbcf2e505bf2f9dd0dc25de4582115ee4ed4a6e80997affc7b22122912cc6591R5470 I think it's best to copy that deprecation.

having `register_worker_plugin` take a class and kwargs separately seems to imply that the class will be constructed somewhere else at another time - rather than immediately

File: distributed/diagnostics/tests/test_worker_plugin.py [hunk 3]
```diff
@@ -84,18 +84,40 @@ async def test_remove_with_client_raises(c, s):
 
 
 @gen_cluster(client=True, nthreads=[])
-async def test_create_with_client_and_plugin_from_class(c, s):
-    await c.register_worker_plugin(MyPlugin, data=456)
+async def test_create_with_client_and_plugin_from_kwargs(c, s, plugin):
```
Comment (graingert) [inline distributed/diagnostics/tests/test_worker_plugin.py:87]:
```suggestion
async def test_create_with_client_and_plugin_from_class(c, s, plugin):
```

File: distributed/diagnostics/tests/test_worker_plugin.py [hunk 4]
```diff
@@ -84,18 +84,40 @@ async def test_remove_with_client_raises(c, s):
 
 
 @gen_cluster(client=True, nthreads=[])
-async def test_create_with_client_and_plugin_from_class(c, s):
-    await c.register_worker_plugin(MyPlugin, data=456)
+async def test_create_with_client_and_plugin_from_class(c, s, plugin):
```
Comment (graingert) [inline distributed/diagnostics/tests/test_worker_plugin.py:87]:
```suggestion
 async def test_create_with_client_and_plugin_from_class(c, s):
```

Comment (graingert) [inline distributed/diagnostics/tests/test_worker_plugin.py]:
```suggestion
async def test_create_with_client_and_plugin_from_class(c, s):
```

File: distributed/client.py [hunk 5]
```diff
@@ -4591,16 +4591,17 @@ def register_worker_plugin(self, plugin=None, name=None, nanny=None, **kwargs):
         Parameters
         ----------
         plugin : WorkerPlugin or NannyPlugin
-            The plugin object to register.
+            WorkerPlugin or NannyPlugin instance to register
```
Comment (graingert) [inline distributed/client.py:4594]:
```suggestion
            WorkerPlugin or NannyPlugin instance to register.
```

File: distributed/diagnostics/tests/test_worker_plugin.py [hunk 6]
```diff
@@ -84,18 +84,40 @@ async def test_remove_with_client_raises(c, s):
 
 
 @gen_cluster(client=True, nthreads=[])
-async def test_create_with_client_and_plugin_from_class(c, s):
-    await c.register_worker_plugin(MyPlugin, data=456)
+ async def test_create_with_client_and_plugin_from_class(c, s):
```
Comment (graingert) [inline distributed/diagnostics/tests/test_worker_plugin.py:87]:
```suggestion
async def test_create_with_client_and_plugin_from_class(c, s):
```

File: distributed/client.py (hunk 2 above)
Comment (graingert) [reply distributed/client.py]: `register_scheduler_plugin` is interesting because it's different again, the plugin gets instantiated in the event loop thread: https://github.com/dask/distributed/blob/9a66b719d53303b51034b339a97a9131105e5451/distributed/client.py#L4498-L4500

whereas for the other plugin registration methods the plugin is instantiated in the current thread

Comment (jrbourbeau) [reply distributed/client.py]: Good catch. Being consistent with when the plugin instance is created would be nice. My guess is either convention is fine. I have a slight preference for creating the plugin inside the event loop thread so that users can be more explicit in `await`ing the registration process.

Comment (fjetter) [reply distributed/client.py]: > That's different as it's for methods on the Scheduler, which are much less user-facing than the Client

Well, we're also raising a `FutureWarning` over there so I assume we expect it to be used by someone?

>  I have a slight preference for creating the plugin inside the event loop thread so that users can be more explicit in awaiting the registration process.

I have a preference to do the opposite since we're talking about a loop on Client side here. I don't see a reason why anyone should interact with the client loop at all. Once on the scheduler/worker this is a different story but for the setup there we have appropriate hooks.

When Thomas and I discussed this earlier I was actually thrown off myself for quite some time because I thought this API would only instantiate the class on the remote but that's not the case.

I don't think having two ways to do the same thing is particularly helping with UX, especially not if there is a looming question of "is it executed on the loop or not?" and "is it instantiated on client side or not?"

The bottom line is that I like this change much more than the exception / argument parsing logic we would have in here if we needed to support both use cases. We'd also need much more explicit docs. It also aligns better with the scheduler

File: distributed/client.py [hunk 7]
```diff
@@ -4512,15 +4512,24 @@ def register_scheduler_plugin(self, plugin, name=None, **kwargs):
...
+            deprecated; Arguments passed to the Plugin class (if Plugin is an
             instance kwargs are unused).
 
         """
+        if isinstance(plugin, type):
+            warnings.warn(
+                "Adding plugins by class is deprecated and will be disabled in a "
+                "future release. Please add plugins by instance instead.",
+                category=FutureWarning,
+            )
```
Comment (graingert) [inline distributed/client.py:4529]:
```suggestion
            )
            # note: plugin is constructed in async def _register_scheduler_plugin
```

**Review by gjoseph92 — APPROVED** [review]
LGTM. I am also +1 on having just one way to do things, even though it is a breaking API change. I don't think the current API adds any value (thing you can do with it that you can't do without it), nor even convenience.

Indeed, passing a `NannyPlugin` class (instead of instance) right now will probably behave incorrectly:
https://github.com/dask/distributed/blob/27e467e56e1dea6a186152551a269de180bb593e/distributed/client.py#L4574-L4575

**Review by jrbourbeau — COMMENTED** [review]
Yeah, I'll walk back on my previous push back to the changes here. Looking into our git history it appears support for passing in a plugin class, instead of an instance, was a workaround for a cryptic error message. We can add more explicit docs / error messages if this becomes an issue in the future 👍

**Review by jrbourbeau — APPROVED** [review]
Thanks @graingert -- this is in
